R/logging.R

```diff
@@ -63,6 +63,156 @@ message_wrap <-
     invisible(msg)
   }
 
+# issue cataloger --------------------------------------------------------------
+tune_env <-
+  rlang::new_environment(
+    data = list(progress_env = NULL, progress_active = FALSE, progress_catalog = NULL)
+  )
+
+# determines whether a currently running tuning process uses the cataloger.
+uses_catalog <- function() {
+  isTRUE(tune_env$progress_active && !is_testing())
+}
+
+# determines whether a tuning process is currently active.
+# this function is intended to guard calls to `initialize_catalog()`
+# inside of machinery that is wrapped by other tuning
+# methods. e.g. `tune_bayes()` has an issue cataloger that ought not to be
+# overwritten when fitting over an initial grid.
+catalog_is_active <- function() {
+  tune_env$progress_active
+}
+
+# initializes machinery for the tune catalog inside of an environment.
+# the `env` should be an execution environment that persists throughout the
+# tuning process for a given tuning approach and exits once tuning is completed.
+#' @rdname tune-internal-functions
+#' @export
+initialize_catalog <- function(control, env = rlang::caller_env()) {
+  catalog <-
+    tibble::tibble(
+      type = character(0),
+      note = character(0),
+      n = numeric(0),
+      id = numeric(0)
+    )
+
+  if (!(allow_parallelism(control$allow_par) || is_testing())) {
+    progress_active <- TRUE
+  } else {
+    progress_active <- FALSE
+  }
+
+  rlang::env_bind(tune_env, progress_env = env)
+
+  rlang::env_bind(tune_env, progress_catalog = catalog)
+  withr::defer(
+    rlang::env_bind(tune_env, progress_catalog = NULL),
+    envir = env
+  )
+
+  rlang::env_bind(tune_env, progress_active = progress_active)
+  withr::defer(
+    rlang::env_bind(tune_env, progress_active = FALSE),
+    envir = env
+  )
+
+
+  invisible(NULL)
+}
+
+# given a catalog, summarize errors and warnings in a 1-length glue vector.
+# for use by the progress bar inside of `tune_catalog()`.
+summarize_catalog <- function(catalog) {
+  if (nrow(catalog) == 0) {
+    return("")
+  }
+
+  res <- dplyr::arrange(catalog, id)
+  res <- dplyr::mutate(res, color = dplyr::if_else(type == "warning", list(cli::col_yellow), list(cli::col_red)))
+  res <- dplyr::rowwise(res)
+  res <- dplyr::mutate(res, msg = glue::glue("{color(cli::style_bold(id))}: x{n}"))
+  res <- dplyr::ungroup(res)
+  res <- dplyr::pull(res, msg)
+  res <- glue::glue_collapse(res, sep = "   ")
```

Comment:
Maybe make the number of spaced an argument? If we eventually nest this inside of another function we might want different amounts of whitespace to differentiate nested catalogs?

Reply:
You got it! [2e94742](https://github.com/tidymodels/tune/pull/593/commits/2e94742c45a469776182246da0dcbd3ca9e92db6)